### PR TITLE
Product is included if Event's taxon is an ancestor of any of its taxons

### PR DESCRIPTION
**Pull request by nando**
Or the other way around, Event will include all products related with its taxon or with any of its descendants.

### CODE REVIEW

File: app/models/spree/product_decorator.rb [hunk 1]
```diff
@@ -2,10 +2,10 @@
   has_many :active_sale_events, :as => :eventable
 
   def find_live_taxons
-    Spree::ActiveSaleEvent.live_active.where(:eventable_type => "Spree::Taxon", :eventable_id => self.taxons.map(&:id))
+    Spree::ActiveSaleEvent.live_active.where(:eventable_type => "Spree::Taxon", :eventable_id => self.taxons.map{|t| t.self_and_ancestors}.flatten.map(&:id))
```
Comment (suryart) [inline app/models/spree/product_decorator.rb:6]: Any good scenario/reason for why we should look for ancestors, too?

**Comment by dfischer** [comment]
Is this a scenario for something like, a taxon of "Clothing" which includes T-Shirts, Hats, etc?

**Comment by suryart** [comment]
Yes. But, I don't understand this, suppose if you have a taxon: "Men", under which you have another taxon: "Formal Shirts", and for "Formal Shirts" you have some products. Now, if "Formal Shirts" sale is down(i.e not live), but, other sales running under "Men" for example: "Designer Shoes", "Sunglasses" are live. Won't this change consider "Men" sale as live and return true instead of false for "Formal Shirts" sale's product. 

That's what making me to have a re-look at this. And have asked that, what's the scenario here? Maybe @nando has something else on his mind.

**Comment by nando** [comment]
Hi Suryart, excuse me for the delay, for some reason I'm not receiving GitHub notifications :(

Yes, maybe it depends on the way you use your taxonomy, in the way you relate your products with your taxons. In the cases I've seen so far parent taxons are used only to group other taxons and they are not directly related with products. Products are usually related with leaf taxons.

In the scenario you've described, if no products are directly related with "Men" the sale only has sense if products related with "Men" descendants are included. It'd be up to the administrator to avoid this kind of scenarios.

**Comment by suryart** [comment]
OK. I've got your point now. But, you can't assume an administrator(ultimately your user) to run into such a little gotcha. It will screw the work flow. 

I have another proposal to re-architect this and actually I am about to do the same. How about implementing STI for ActiveSale and ActiveSaleEvent and ActiveSale will not hold eventable(i.e. product/ taxon) just like before. Whole workflow will be the same as existing, and ActiveSale's start and end date will depend on the active_sale_events with in it also it will have parent child relationships. 

For example:

An ActiveSale name: "Men" can have another ActiveSale in it: "Apparel" which can then have another ActiveSale(if required): "Shirts" which will hold active sale events. Each event can be a product or a taxon. 

Here the trick is: If there are 3 events("A - start_date: 2013-03-08 10:43:57 end_date: 2013-03-18 10:43:57", "B - start_date: 2013-03-04 10:43:57 end_date: 2013-03-13 10:43:57", and "C - start_date: 2013-03-15 10:43:57 end_date: 2013-03-28 10:43:57") in Shirts then its parents active_sales' start and date will be expanded accordingly(i.e. Shirt's active sale's start_date will be: 2013-03-04 10:43:57 and end_date will be: 2013-03-28 10:43:57) for the sake of being live using active record callbacks. 

This gives less sql queries to check if a sale is live. Second you can generate navigation on the store dynamically without iterating and querying more data from sql for events.

Let me know what do you think?

Also, can you please update your repository to the latest?

**Comment by nando** [comment]
Ooops! I've broken the build :( I'll try to fix it ASAP. Forget my current repo, I'll recreate it from your current master and create a new branch for this issue.

Your new architect sounds a little bit complex to me, I like a lot the simplicity of the current model... but I'd like to think a little bit more about it to understand your motivations (queries, navigation, etc.).

To resolve my scenario I'd simply add a checkbox "Include descendants" next to the taxon name if the eventable is a taxon.

**Comment by suryart** [comment]
Well, mechanism is a bit complex. But, it will hit many birds from one stone. 

I'll create an another branch for it. So, everybody can look at it and see if it's good or not.

**Comment by nando** [comment]
Nice, good idea.

I've splitted my changes in two different branches: one for this issue (https://github.com/suryart/spree_active_sale/pull/17) and other for the spanish translations, both from your current master and with a green build :)

I'll close this issue keeping a link to it from the new one.

**Pull request closed without merging**